File: sbr android/app/src/main/java/com/sbr/sms/ui/customer/viewmodels/ReferralViewModel.kt
package com.sbr.sms.ui.customer.viewmodels

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sbr.sms.data.api.ProductDto
import com.sbr.sms.data.api.ReferralClaimDto
import com.sbr.sms.data.api.ReferralDashboardDto
import com.sbr.sms.data.api.ReferralDto
import com.sbr.sms.data.api.SubmitReferralRequest
import com.sbr.sms.data.api.ClaimPayoutRequest
import com.sbr.sms.data.repositories.ReferralRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ReferralUiState(
    val isLoading: Boolean = false,
    val error: String? = null,
    val dashboard: ReferralDashboardDto? = null,
    val products: List<ProductDto> = emptyList(),
    val isSubmittingReferral: Boolean = false,
    val referralSuccessMsg: String? = null,
    val referralErrorMsg: String? = null,
    val isSubmittingClaim: Boolean = false,
    val claimSuccessMsg: String? = null,
    val claimErrorMsg: String? = null
)

@HiltViewModel
class ReferralViewModel @Inject constructor(
    private val repository: ReferralRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(ReferralUiState())
    val uiState: StateFlow<ReferralUiState> = _uiState.asStateFlow()

    init {
        loadData()
    }

    fun loadData() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            
            val dashboardResult = repository.getMyReferrals()
            val productsResult = repository.getProducts()

            if (dashboardResult.isSuccess) {
                val dashboardData = dashboardResult.getOrNull()
                val productsData = productsResult.getOrDefault(emptyList())
                _uiState.update { 
                    it.copy(
                        isLoading = false,
                        dashboard = dashboardData,
                        products = productsData
                    )
                }
            } else {
                _uiState.update { 
                    it.copy(
                        isLoading = false,
                        error = dashboardResult.exceptionOrNull()?.message ?: "Failed to load referral details"
                    )
                }
            }
        }
    }

    fun submitReferral(refereeName: String, refereePhone: String, productId: String?, productName: String, notes: String?) {
        viewModelScope.launch {
            _uiState.update { it.copy(isSubmittingReferral = true, referralSuccessMsg = null, referralErrorMsg = null) }
            val req = SubmitReferralRequest(refereeName, refereePhone, productId, productName, notes)
            val result = repository.submitReferral(req)
            if (result.isSuccess) {
                _uiState.update { 
                    it.copy(
                        isSubmittingReferral = false,
                        referralSuccessMsg = "Referral lead submitted successfully!"
                    )
                }
                loadData() // Refresh dashboard stats and lists
            } else {
                _uiState.update { 
                    it.copy(
                        isSubmittingReferral = false,
                        referralErrorMsg = result.exceptionOrNull()?.message ?: "Failed to submit referral"
                    )
                }
            }
        }
    }

    fun claimPayout(amount: Double, payoutMethod: String, payoutDetails: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isSubmittingClaim = true, claimSuccessMsg = null, claimErrorMsg = null) }
            val req = ClaimPayoutRequest(amount, payoutMethod, payoutDetails)
            val result = repository.claimPayout(req)
            if (result.isSuccess) {
                _uiState.update { 
                    it.copy(
                        isSubmittingClaim = false,
                        claimSuccessMsg = "Payout claim request submitted successfully!"
                    )
                }
                loadData() // Refresh dashboard stats and lists
            } else {
                _uiState.update { 
                    it.copy(
                        isSubmittingClaim = false,
                        claimErrorMsg = result.exceptionOrNull()?.message ?: "Failed to claim payout"
                    )
                }
            }
        }
    }

    fun clearMessages() {
        _uiState.update { 
            it.copy(
                referralSuccessMsg = null,
                referralErrorMsg = null,
                claimSuccessMsg = null,
                claimErrorMsg = null
            )
        }
    }
}

File: sbr ios/SBR/SBR/ViewModels/ReferralViewModel.swift
import Foundation
import Combine

@MainActor
class ReferralViewModel: ObservableObject {
    @Published var dashboard: ReferralDashboard? = nil
    @Published var allReferrals: [Referral] = []
    @Published var allClaims: [ReferralClaim] = []
    @Published var products: [Product] = [] // For selecting products in referral form
    @Published var isLoading = false
    @Published var successMessage: String? = nil
    @Published var errorMessage: String? = nil
    
    func clearMessages() {
        successMessage = nil
        errorMessage = nil
    }
    
    // CUSTOMER METHODS
    func fetchMyReferrals() async {
        isLoading = true
        errorMessage = nil
        do {
            let res = try await APIClient.shared.get(endpoint: "api/referrals/my-referrals", responseType: APIResponse<ReferralDashboard>.self)
            if res.success, let data = res.data {
                self.dashboard = data
            } else {
                self.errorMessage = res.error ?? "Failed to fetch referral dashboard"
            }
        } catch {
            self.errorMessage = error.localizedDescription
        }
        isLoading = false
    }
    
    func fetchProductsForReferral() async {
        do {
            // Load products list for dropdown
            let res = try await APIClient.shared.get(endpoint: "api/products?activeOnly=false", responseType: APIResponse<[Product]>.self)
            if res.success, let data = res.data {
                self.products = data
            }
        } catch {
            print("Failed to fetch products for referral dropdown: \(error)")
        }
    }
    
    func submitReferral(name: String, phone: String, productId: String?, productName: String, notes: String?) async -> Bool {
        isLoading = true
        errorMessage = nil
        successMessage = nil
        let req = SubmitReferralRequest(refereeName: name, refereePhone: phone, productId: productId, productName: productName, notes: notes)
        do {
            let res = try await APIClient.shared.post(endpoint: "api/referrals/submit", body: req, responseType: APIResponse<Referral>.self)
            if res.success {
                self.successMessage = "Referral submitted successfully!"
                await fetchMyReferrals()
                isLoading = false
                return true
            } else {
                self.errorMessage = res.error ?? "Failed to submit referral"
            }
        } catch {
            self.errorMessage = error.localizedDescription
        }
        isLoading = false
        return false
    }
    
    func claimPayout(amount: Double, method: String, details: String) async -> Bool {
        isLoading = true
        errorMessage = nil
        successMessage = nil
        let req = ClaimPayoutRequest(amount: amount, payoutMethod: method, payoutDetails: details)
        do {
            let res = try await APIClient.shared.post(endpoint: "api/referrals/claim-payout", body: req, responseType: APIResponse<ReferralClaim>.self)
            if res.success {
                self.successMessage = "Payout request submitted successfully!"
                await fetchMyReferrals()
                isLoading = false
                return true
            } else {
                self.errorMessage = res.error ?? "Failed to request payout"
            }
        } catch {
            self.errorMessage = error.localizedDescription
        }
        isLoading = false
        return false
    }
    
    // ADMIN METHODS
    func adminFetchAllReferrals() async {
        isLoading = true
        errorMessage = nil
        do {
            let res = try await APIClient.shared.get(endpoint: "api/referrals/admin/all", responseType: APIResponse<[Referral]>.self)
            if res.success, let data = res.data {
                self.allReferrals = data
            } else {
                self.errorMessage = res.error ?? "Failed to fetch referrals"
            }
        } catch {
            self.errorMessage = error.localizedDescription
        }
        isLoading = false
    }
    
    func adminFetchAllClaims() async {
        isLoading = true
        errorMessage = nil
        do {
            let res = try await APIClient.shared.get(endpoint: "api/referrals/admin/claims", responseType: APIResponse<[ReferralClaim]>.self)
            if res.success, let data = res.data {
                self.allClaims = data
            } else {
                self.errorMessage = res.error ?? "Failed to fetch payout claims"
            }
        } catch {
            self.errorMessage = error.localizedDescription
        }
        isLoading = false
    }
    
    func adminUpdateReferralStatus(id: String, status: String, purchaseAmount: Double?, rewardAmount: Double?, notes: String?) async -> Bool {
        isLoading = true
        errorMessage = nil
        successMessage = nil
        let req = UpdateReferralStatusRequest(status: status, purchaseAmount: purchaseAmount, rewardAmount: rewardAmount, notes: notes)
        do {
            let res = try await APIClient.shared.put(endpoint: "api/referrals/admin/\(id)/status", body: req, responseType: APIResponse<Referral>.self)
            if res.success {
                self.successMessage = "Referral updated successfully!"
                await adminFetchAllReferrals()
                isLoading = false
                return true
            } else {
                self.errorMessage = res.error ?? "Failed to update referral"
            }
        } catch {
            self.errorMessage = error.localizedDescription
        }
        isLoading = false
        return false
    }
    
    func adminUpdateClaimStatus(id: String, status: String, transactionRef: String?, adminNotes: String?) async -> Bool {
        isLoading = true
        errorMessage = nil
        successMessage = nil
        let req = UpdateClaimStatusRequest(status: status, transactionRef: transactionRef, adminNotes: adminNotes)
        do {
            let res = try await APIClient.shared.put(endpoint: "api/referrals/admin/claims/\(id)/status", body: req, responseType: APIResponse<ReferralClaim>.self)
            if res.success {
                self.successMessage = "Claim status updated successfully!"
                await adminFetchAllClaims()
                isLoading = false
                return true
            } else {
                self.errorMessage = res.error ?? "Failed to update claim"
            }
        } catch {
            self.errorMessage = error.localizedDescription
        }
        isLoading = false
        return false
    }
}
